presto-main/src/main/java/io/prestosql/server/DynamicFilterService.java

```diff
@@ -0,0 +1,268 @@
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+import static io.airlift.concurrent.MoreFutures.addSuccessCallback;
+import static io.airlift.concurrent.Threads.daemonThreadsNamed;
+import static io.prestosql.execution.StageState.FINISHED;
+import static io.prestosql.execution.StageState.RUNNING;
+import static io.prestosql.execution.StageState.SCHEDULED;
+import static io.prestosql.execution.StageState.SCHEDULING_SPLITS;
+import static io.prestosql.spi.predicate.TupleDomain.columnWiseUnion;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newSingleThreadScheduledExecutor;
+import static java.util.stream.Collectors.groupingBy;
+import static java.util.stream.Collectors.mapping;
+
+@ThreadSafe
+public class DynamicFilterService
+{
+    private final ConcurrentHashMap<SourceDescriptor, SettableFuture<TupleDomain<String>>> dynamicFilterSummaries = new ConcurrentHashMap<>();
+    private final ConcurrentMap<QueryId, Supplier<List<StageInfo>>> queries = new ConcurrentHashMap<>();
+    private final ScheduledExecutorService collectDynamicFiltersExecutor = newSingleThreadScheduledExecutor(daemonThreadsNamed("DynamicFilterService"));
+
+    @Inject
+    public DynamicFilterService(TaskManagerConfig taskConfig)
+    {
+        collectDynamicFiltersExecutor.scheduleWithFixedDelay(this::collectDynamicFilters, 0, taskConfig.getStatusRefreshMaxWait().toMillis(), TimeUnit.MILLISECONDS);
+    }
+
+    public void registerQuery(QueryId queryId, Supplier<List<StageInfo>> stageInfoSupplier)
+    {
+        queries.putIfAbsent(queryId, stageInfoSupplier);
+    }
+
+    @VisibleForTesting
+    public void collectDynamicFilters()
+    {
+        List<QueryId> queriesWithDynamicFilters = dynamicFilterSummaries.entrySet()
+                .stream()
+                .filter(entry -> !entry.getValue().isDone())
+                .map(entry -> entry.getKey().getQueryId())
+                .distinct()
+                .collect(toImmutableList());
+
+        for (QueryId queryId : queriesWithDynamicFilters) {
+            Supplier<List<StageInfo>> stageInfoSupplier = queries.get(queryId);
+            if (stageInfoSupplier != null) {
+                for (StageInfo stageInfo : getStagesWithDynamicFilters(stageInfoSupplier.get())) {
+                    StageState stageState = stageInfo.getState();
+                    // wait until stage has finished scheduling tasks
+                    if ((stageState == SCHEDULING_SPLITS || stageState == SCHEDULED || stageState == RUNNING || stageState == FINISHED)) {
```

Comment:
revert the condition:
```
if (!(stageState == SCHEDULING_SPLITS || stageState == SCHEDULED || stageState == RUNNING || stageState == FINISHED)) {
  continue
}
```
actually, you could extract this condition to separate method or maybe add another field (`noMoreTasks`) to `io.prestosql.execution.StageState`. This way you could also simplify `io.prestosql.execution.scheduler.SqlQueryScheduler.StageLinkage#processScheduleResults`

@dain Are we guaranteed that there will be no more tasks added (e.g when new workers come with co-located joins) when we reach `SCHEDULING_SPLITS` state?

Reply:
I've extracted the logic from `StageLinkage#processScheduleResults` into a method in `StageState` and reused it.